Fix test failures for release / tarball builds

jamessan opened the pull request:
Now that I've finally got the new dependencies packaged for 0.10.x, I've started running into test failures due to expectations of running with a debug build or from a git clone.

I'll adjust the CI to catch these problems and add fixes.

jamessan committed:
ci: run one set of tests with a release build

This ensures that no tests fail due to differences between release and debug builds.

The release build-type check is now unnecessary, too, so remove it.

jamessan committed:
ci(test): remove the .git directory for Linux

Tests should not rely on being run inside a git clone, so the Linux
builds cover this use case. The macOS builds will continue running with
the .git directory so there's still unix-ish coverage within a git
clone.

dundargoc commented on .github/workflows/test.yml:
```diff
@@ -145,6 +145,9 @@ jobs:
           perl -W -e 'use Neovim::Ext; print $Neovim::Ext::VERSION'
 
+      - name: Remove .git directory
+        run: cmake -E rm -rf -- .git
```
maybe just fix the three tests that fail instead of encoding this for eternity?

dundargoc commented on .github/workflows/test.yml:
```diff
@@ -145,6 +145,9 @@ jobs:
           perl -W -e 'use Neovim::Ext; print $Neovim::Ext::VERSION'
 
+      - name: Remove .git directory
+        run: cmake -E rm -rf -- .git
```
This is an expensive fix for a problem that probably won't happen again unless we create another root traversal function. I think a better tradeoff would just be to fix the tests and call it a day.

jamessan replied:
I'm going to fix the tests. This just helps ensure there aren't other things relying on git (like the build part of the version inf). I'm not sure what's expensive about it. It takes less than a second in CI.

dundargoc replied:
I meant more the maintaining this is expensive compared to the problem, which probably won't occur again.

jamessan committed:
test(main_spec): use CMakePresets.json instead of .git for root marker

jamessan committed:
test(version_spec): expect vim.NIL, not nil, for "build" if not in a git clone

jamessan committed:
test(main_spec): make "nvim -v" test agnostic to build type

In release builds, the Compilation: line is omitted so the build is reproducible. Since the "fall-back for $VIM" line is always present, check for that instead.

dundargoc commented on .github/workflows/test.yml:
```diff
@@ -145,14 +146,22 @@ jobs:
 
+      - if: ${{ matrix.build.flavor == 'release' }}
+        name: Setup Release build type
+        run: echo CMAKE_BUILD_TYPE=Release >> $GITHUB_ENV
```
Wut? Does adding `CMAKE_BUILD_TYPE=Release` to flags not work?

dundargoc replied:
I checked locally and I can't reproduce. Have you tested it out?

dundargoc replied:
We need to fix that then, this is pretty hacky. We can use this workaround for now in any case.

jamessan replied:
We can touch the respective files earlier in the workflow. The root of the problem is we're using cmake to build and make to test.

dundargoc replied:
Yes, that was mostly to save a ci step since we can't run the oldtests with cmake. I guess it's better to just run tests directly with cmake instead anyway?

jamessan committed:
ci: run tests directly rather than via the Makefile

Since the Makefile is not used to build, running the tests via the Makefile causes cmake to reconfigure and revert the release build back to debug.

zeertzjq commented on .github/workflows/test.yml:
```diff
@@ -194,7 +204,7 @@ jobs:
 
   # This job tests the following things:
-  # - Check if Release, MinSizeRel and RelWithDebInfo compiles correctly.
+  # - Check if  MinSizeRel and RelWithDebInfo compiles correctly.
```
```suggestion
  # - Check if MinSizeRel and RelWithDebInfo compiles correctly.
```

jamessan committed:
test(marktree): expose test functions in release builds

In order to run the marktree unit test in release mode, the test functions need to be available even when NDEBUG is defined.

Keep the body of marktree_check a nop during release builds, which limits the usefulness of the testing, but at least lets the tests run.

jamessan commented:
N.B., this changes the name of the "ubuntu gcc ..." jobs (which are "required"), so that will need to be adjusted as part of this merge.

justinmk commented:
> this changes the name of the "ubuntu gcc ..." jobs (which are "required"), so that will need to be adjusted as part of this merge.

good call out , but I don't see any place where we actually have enumerated required status checks.

<img width="418" alt="image" src="https://github.com/user-attachments/assets/bad84eac-2faf-46af-b758-a43f130274ca">
